core/metacling/src/TClingCallFunc.cxx

```diff
@@ -383,9 +383,23 @@ void TClingCallFunc::make_narg_call(const std::string &return_type, const unsign
          callbuf << "*(" << type_name.c_str() << "**)args["
                  << i << "]";
       } else {
-         // pointer falls back to non-pointer case; the argument preserves
-         // the "pointerness" (i.e. doesn't reference the value).
-         callbuf << "*(" << type_name.c_str() << "*)args[" << i << "]";
+         // By-value construction; this may either copy or move, but there is no
+         // information here in terms of intent. Thus, simply assume that the intent
+         // is to move if there is no viable copy constructor (ie. if the code would
+         // otherwise fail to even compile).
+
+         // There does not appear to be a simple way of determining whether a viable
+         // copy constructor exists, so check for the most common case: the trivial
+         // one, but not uniquely available, while there is a move constructor.
+         CXXRecordDecl* rtdecl = QT->getAsCXXRecordDecl();
+         if (rtdecl && (rtdecl->hasTrivialCopyConstructor() && !rtdecl->hasSimpleCopyConstructor()) \
```

Comment:
Why require `rtdecl->hasTrivialCopyConstructor()`? AFAICT it should be sufficient to check `!rtdecl->hasSimpleCopyConstructor()`... (what does `rt` stand for? There is no `t` in `CXXRecordDecl`... IIRC the usual acronym is `RD` in upstream Clang code)

Also, please remove the final `\`, the line isn't in a macro...

Reply:
I wanted to check if the changes from cppyy worked as-is, it seems like they do (except for Windows for now). Your suggestions are very valid and I will include them in the next changes. But I see no complaint about the overall direction of the PR, so I'm glad about that!

Reply:
We will probably need a unittest for callfunc as that’s a critical component for root.

Reply:
Ok, I'll try to come up with something that is equivalent to what PyROOT tries to do in this case.

Reply:
I have included a first test, based on already available tests for TClingCallFunc. Let me know what you think

Reply:
It seems that the currently failing tests (e.g. [here](https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/193543/testReport/projectroot.roottest.python/cpp/roottest_python_cpp_advanced/)) are due to the missing `hasTrivialCopyConstructor` check, I will reintroduce it in a new commit

Reply:
Whatever we do we should do it in sync with upstream otherwise it will be a nightmare to upgrade. If the patch differs we should propose it upstream, if it is only the test we should propose the test. That investment will make our lives easier in near future.

Reply:
In principle we can pick what upstream does literally as-is (if we can afford living with the `\` character in the if condition). It is not only the test clearly, as this feature was never present before in ROOT's TClingCallFunc. The main issue I see with picking upstream as-is is that there is no way to pinpoint the necessary commits, the whole TClingCallFunc::make_narg_call belongs to a single commit when the fork of ROOT was started https://github.com/wlav/cppyy-backend/blame/25caf988cef1f2f76705c07b7262f076e8ed0e01/cling/src/core/metacling/src/TClingCallFunc.cxx#L478 . I have no problem with copying upstream code verbatim though. Let me know what I should do because this problem is quite visible in experiment code using PyROOT.